diff -ruN pintos1/src/lib/syscall-nr.h pintos2/src/lib/syscall-nr.h
--- pintos1/src/lib/syscall-nr.h	2022-03-02 06:23:26.000000000 +0900
+++ pintos2/src/lib/syscall-nr.h	2022-03-23 23:32:11.000000000 +0900
@@ -18,6 +18,9 @@
     SYS_SEEK,                   /* Change position in a file. */
     SYS_TELL,                   /* Report current position in a file. */
     SYS_CLOSE,                  /* Close a file. */
+    SYS_SIGACTION,              /* Register an signal handler */
+    SYS_SENDSIG,                /* Send a signal */
+    SYS_YIELD,                  /* Yield current thread */
 
     /* Project 3 and optionally project 4. */
     SYS_MMAP,                   /* Map a file into memory. */
diff -ruN pintos1/src/lib/user/syscall.c pintos2/src/lib/user/syscall.c
--- pintos1/src/lib/user/syscall.c	2022-03-02 06:23:26.000000000 +0900
+++ pintos2/src/lib/user/syscall.c	2022-03-23 23:29:57.000000000 +0900
@@ -141,6 +141,21 @@
   syscall1 (SYS_CLOSE, fd);
 }
 
+void sigaction (int signum, void (*handler) (void))
+{
+  syscall2 (SYS_SIGACTION, signum, handler);
+}
+
+void sendsig (pid_t pid, int signum)
+{
+  syscall2 (SYS_SENDSIG, pid, signum);
+}
+
+void sched_yield ()
+{
+  syscall0 (SYS_YIELD);
+}
+
 mapid_t
 mmap (int fd, void *addr)
 {
diff -ruN pintos1/src/lib/user/syscall.h pintos2/src/lib/user/syscall.h
--- pintos1/src/lib/user/syscall.h	2022-03-02 06:23:26.000000000 +0900
+++ pintos2/src/lib/user/syscall.h	2022-03-23 23:29:57.000000000 +0900
@@ -33,6 +33,12 @@
 void seek (int fd, unsigned position);
 unsigned tell (int fd);
 void close (int fd);
+void sigaction (int signum, void (*handler) (void));
+void sendsig (pid_t, int signum);
+#define SIGONE 1
+#define SIGTWO 2
+#define SIGTHREE 3
+
 
 /* Project 3 and optionally project 4. */
 mapid_t mmap (int fd, void *addr);
diff -ruN pintos1/src/tests/userprog/child-sig.c pintos2/src/tests/userprog/child-sig.c
--- pintos1/src/tests/userprog/child-sig.c	1970-01-01 09:00:00.000000000 +0900
+++ pintos2/src/tests/userprog/child-sig.c	2022-03-23 23:28:16.000000000 +0900
@@ -0,0 +1,28 @@
+/* Child process run by exec-multiple, exec-one, wait-simple, and
+   wait-twice tests.
+   Just prints a single message and terminates. */
+
+#include <stdio.h>
+#include <string.h>
+#include <syscall.h>
+
+#include "tests/lib.h"
+
+void sig_handler1(void);
+void sig_handler2(void);
+
+int main(void) {
+    test_name = "child-sig";
+
+    sigaction(SIGONE, sig_handler1);
+    sigaction(SIGTWO, sig_handler2);
+    msg("run");
+    sched_yield();
+    return 81;
+}
+
+void sig_handler1(void) {
+}
+
+void sig_handler2(void) {
+}
\ No newline at end of file
diff -ruN pintos1/src/tests/userprog/Grading pintos2/src/tests/userprog/Grading
--- pintos1/src/tests/userprog/Grading	2022-03-02 06:23:26.000000000 +0900
+++ pintos2/src/tests/userprog/Grading	2022-03-24 00:02:33.000000000 +0900
@@ -5,7 +5,8 @@
 # If you do so properly, the base file system functionality
 # should come "for free".  Thus, the points emphasis below.
 
-35%	tests/userprog/Rubric.functionality
-25%	tests/userprog/Rubric.robustness
-10%	tests/userprog/no-vm/Rubric
-30%	tests/filesys/base/Rubric
+24.5%	tests/userprog/Rubric.functionality
+17.5%	tests/userprog/Rubric.robustness
+30%	tests/userprog/Rubric.signal
+7%	tests/userprog/no-vm/Rubric
+21%	tests/filesys/base/Rubric
diff -ruN pintos1/src/tests/userprog/Make.tests pintos2/src/tests/userprog/Make.tests
--- pintos1/src/tests/userprog/Make.tests	2022-03-02 06:23:26.000000000 +0900
+++ pintos2/src/tests/userprog/Make.tests	2022-03-23 23:28:26.000000000 +0900
@@ -16,10 +16,10 @@
 exec-bound-3 exec-multiple exec-missing exec-bad-ptr wait-simple        \
 wait-twice wait-killed wait-bad-pid multi-recurse multi-child-fd        \
 rox-simple rox-child rox-multichild bad-read bad-write bad-read2        \
-bad-write2 bad-jump bad-jump2)
+bad-write2 bad-jump bad-jump2 sig-simple)
 
 tests/userprog_PROGS = $(tests/userprog_TESTS) $(addprefix \
-tests/userprog/,child-simple child-args child-bad child-close child-rox)
+tests/userprog/,child-simple child-args child-bad child-close child-rox child-sig)
 
 tests/userprog/args-none_SRC = tests/userprog/args.c
 tests/userprog/args-single_SRC = tests/userprog/args.c
@@ -100,12 +100,14 @@
 tests/userprog/rox-child_SRC = tests/userprog/rox-child.c tests/main.c
 tests/userprog/rox-multichild_SRC = tests/userprog/rox-multichild.c	\
 tests/main.c
+tests/userprog/sig-simple_SRC = tests/userprog/sig-simple.c tests/main.c
 
 tests/userprog/child-simple_SRC = tests/userprog/child-simple.c
 tests/userprog/child-args_SRC = tests/userprog/args.c
 tests/userprog/child-bad_SRC = tests/userprog/child-bad.c tests/main.c
 tests/userprog/child-close_SRC = tests/userprog/child-close.c
 tests/userprog/child-rox_SRC = tests/userprog/child-rox.c
+tests/userprog/child-sig_SRC = tests/userprog/child-sig.c
 
 $(foreach prog,$(tests/userprog_PROGS),$(eval $(prog)_SRC += tests/lib.c))
 
@@ -134,6 +136,7 @@
 tests/userprog/exec-multiple_PUTFILES += tests/userprog/child-simple
 tests/userprog/wait-simple_PUTFILES += tests/userprog/child-simple
 tests/userprog/wait-twice_PUTFILES += tests/userprog/child-simple
+tests/userprog/sig-simple_PUTFILES += tests/userprog/child-sig
 
 tests/userprog/exec-arg_PUTFILES += tests/userprog/child-args
 tests/userprog/exec-bound_PUTFILES += tests/userprog/child-args
diff -ruN pintos1/src/tests/userprog/Rubric.signal pintos2/src/tests/userprog/Rubric.signal
--- pintos1/src/tests/userprog/Rubric.signal	1970-01-01 09:00:00.000000000 +0900
+++ pintos2/src/tests/userprog/Rubric.signal	2022-03-23 23:28:47.000000000 +0900
@@ -0,0 +1,3 @@
+Signal implementation:
+- Test signal functionality.
+3	sig-simple
diff -ruN pintos1/src/tests/userprog/sig-simple.c pintos2/src/tests/userprog/sig-simple.c
--- pintos1/src/tests/userprog/sig-simple.c	1970-01-01 09:00:00.000000000 +0900
+++ pintos2/src/tests/userprog/sig-simple.c	2022-03-23 23:28:58.000000000 +0900
@@ -0,0 +1,17 @@
+/* Executes and waits for a single child process. */
+
+#include <syscall.h>
+#include "tests/lib.h"
+#include "tests/main.h"
+
+void sig_handler (void);
+
+void
+test_main (void) 
+{
+  pid_t child_pid = exec ("child-sig");
+  sendsig (child_pid, SIGONE);
+  sendsig (child_pid, SIGTWO);
+  sendsig (child_pid, SIGTHREE);
+  wait (child_pid);
+}
diff -ruN pintos1/src/tests/userprog/sig-simple.ck pintos2/src/tests/userprog/sig-simple.ck
--- pintos1/src/tests/userprog/sig-simple.ck	1970-01-01 09:00:00.000000000 +0900
+++ pintos2/src/tests/userprog/sig-simple.ck	2022-03-23 23:28:58.000000000 +0900
@@ -0,0 +1,14 @@
+# -*- perl -*-
+use strict;
+use warnings;
+use tests::tests;
+check_expected ([<<'EOF']);
+(sig-simple) begin
+(child-sig) run
+Signum: 1, Action: 0x80480a0
+Signum: 2, Action: 0x80480a1
+child-sig: exit(81)
+(sig-simple) end
+sig-simple: exit(0)
+EOF
+pass;
